document-portal/document-portal.c

```diff
@@ -435,7 +433,63 @@ validate_fd_common (int fd,
   if (st_mode == S_IFDIR)
     required_mode |= X_OK;
 
-  access_mode = get_fd_access (fd);
+  if ((fd_flags & O_PATH) == O_PATH)
+    {
+      /* Earlier versions of the portal supported only O_PATH fds, as
+       * these are safer to handle on the portal side. But we now
+       * prefer regular FDs because these ensure that the sandbox
+       * actually has full access to the file in its security context.
+       *
+       * However, we still support O_PATH fds when possible because
+       * existing code uses it, and it is safe for flatpak sandboxes
+       * since they run in the same security context as the portal.
+       *
+       * See issues #167 for details.
+       */
+
+      /* For O_PATH we know only that we can see the path, not that we
+       * can access it, which could be different in the sandboxs
+       * compared to the host. For example, the sandbox can have the
+       * fs mounted readonly. We do this check with access on the
+       * sandbox supplied fd, as that applies the mount flags from the
+       * sandbox.
+       *
+       * Note however that the access() call doesn't apply other parts
+       * of the security context, like AppArmour, so this only works
+       * if we don't realy on that.
+       */
+
+      if (xdp_app_info_supports_opath (app_info))
+        access_mode = get_fd_access (fd);
+      else
+        access_mode = -1;
+    }
+  else if (st_mode == S_IFDIR)
+    {
+      /* For directories we only support O_PATH, which is fine because
+       * they are only used by AddNamed* which is allowed from the host only.
+       */
```

Comment:
I'm not sure this is correct.  If I want to open a folder in Nautilus using `OpenURI.OpenFile`, then we'd have a directory descriptor here.

We'd only need to prove read access for that case, which we should be able to handle with non-O_PATH descriptors.  Given that you can't open an O_RDWR or O_WRONLY descriptor for a directory, perhaps the logic for regular files would be sufficient for directories too?

Reply:
OpenURI.OpenFile uses a different codepath (xdp_app_info_get_path_for_fd), so this comment doesn't affect this change. However, it is true that we want that to support non-O_PATH fds too. I'll have a look at that.